### PR TITLE
per tenant remote sync properly handle metadata update

### PR DESCRIPTION
**Pull request by LizardWizzard**
See commit messages for details

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -431,11 +479,29 @@ impl Tenant {
...
+                    remote_startup_data.as_ref().map(|r| &r.index_part),
+                )
                 .await
                 .context("failed to reconcile with remote")?
         }
 
+        // Save the metadata file to local disk.
+        // Do this last, so that if we crash in-between, we won't think that the
+        // local state is valid.
+        // FIXME, we wont anyway since we have attach mark file. Can still do this at the end
```
Comment (LizardWizzard) [inline pageserver/src/tenant.rs:493]: Should we change this, or leave as is? Any objections to leave as is?

Comment (hlinnaka) [reply pageserver/src/tenant.rs:493]: What would be the alternative? Update the metadata file earlier already? 

Seems fine as it is.

**Review by problame — COMMENTED** [review]
LGTM.

I wasted some time trying to avoid the

```
(None, None) => anyhow::bail!("we should have either local metadata or remote"),
```

case by turning RemoteSetupData into an enum that distinguishes the three other cases

```
enum RemoteSetupData {
  LocalOnly { local_metadata}
  RemoteOnly { index_part }
  Both { local_metadata, index_part }
}
```

But the end result wasn't much better, so, the increased diff isn't justified.

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -855,8 +925,33 @@ impl Tenant {
...
+        let remote_startup_data = match &remote_client {
+            Some(remote_client) => match remote_client.download_index_file().await {
+                Ok(index_part) => {
+                    let remote_metadata = index_part.parse_metadata().context("parse_metadata")?;
+                    Some(RemoteStartupData {
+                        index_part,
+                        remote_metadata,
+                    })
+                }
+                Err(DownloadError::NotFound) => {
```
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:937]: Yeah, we get it from NoSuchKey here  https://github.com/neondatabase/neon/blob/main/libs/remote_storage/src/s3_bucket.rs#L218

**Pull request merged**